homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,230 @@
+
+                hass.loop.call_later(reconnect_interval, reconnect, exc)
+                return
+
+            # Load platforms
+            for comp_name, comp_conf in switches.items():
+                hass.async_create_task(
+                    async_load_platform(hass, 'switch', DOMAIN,
+                                        (comp_name, comp_conf, device),
+                                        config))
+
+            # Bind protocol to command class to allow entities to send commands
+            SW16Device.set_hlk_sw16_protocol(protocol, device)
+
+            # handle shutdown of HLK-SW16 asyncio transport
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                       lambda x: transport.close())
+
+            _LOGGER.info('Connected to HLK-SW16')
+
+        hass.async_create_task(connect())
+
+    for device in config[DOMAIN]:
+        add_device(device)
+    return True
+
+
+class SW16Device(Entity):
+    """Representation of a HLK-SW16 device.
+
+    Contains the common logic for HLK-SW16 entities.
+    """
+
+    platform = None
+    _available = True
+    _protocol_reg = {}
+
+    def __init__(self, relay_name, device_port, device_id, name=None):
+        """Initialize the device."""
+        # HLK-SW16 specific attributes for every component type
+        self._device_id = device_id
+        self._device_port = device_port
+        self._is_on = None
+        self._protocol = self._protocol_reg[device_id]
+        self._name = relay_name
+
+    @classmethod
+    def set_hlk_sw16_protocol(cls, protocol, device_id):
+        """Set the HLK-S16 asyncio protocol as a class variable."""
+        cls._protocol_reg[device_id] = protocol
+
+    def is_connected(self):
+        """Return connection status."""
+        return bool(self._protocol)
+
+    @callback
+    def handle_event_callback(self, event):
+        """Propagate changes through ha."""
+        _LOGGER.debug("Relay %s new state callback: %r",
+                      self._device_port, event)
+        self._is_on = event
+        self.async_schedule_update_ha_state()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._is_on
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @callback
+    def _availability_callback(self, availability):
+        """Update availability state."""
+        self._available = availability
+        self.async_schedule_update_ha_state()
+
+    async def async_added_to_hass(self):
+        """Register update callback."""
+        self._protocol.register_status_callback(self.handle_event_callback,
```

Comment:
Connect the protocol set method here in async_added_to_hass somewhere.

The protocol set method should be signaled after loss of connection when the connection is established again.

Reply:
Is that needed if I'm passing it to __init__ when creating the entity?

Reply:
I thought we needed to update the entity with a new protocol after loosing connection and reconnecting? That's why I proposed the above.

Reply:
Oh, wonder if that could just be handled within `_availability_callback`.

Reply:
Do I also need to namespace `SIGNAL_AVAILABILITY` to a specific hlk-sw16 device?

Reply:
I'm thinking moving the reconnect logic into the protocol itself may make the implementation cleaner, I'm having trouble following the codepaths for the reconnect logic in the component.

Reply:
You mean moving the reconnect logic to the device library?

Reply:
Yeah

Reply:
That's totally fine. We probably should just make sure to listen to home assistant stop event to clean up any remaining reconnection tasks.

Reply:
I should just attach that listener to a new protocol method that permanently shuts it down right?

Reply:
Yes. If you're using asyncio, you probably want to cancel the task that is doing the reconnecting, if that has been scheduled and is not done.

Reply:
Yeah, I'm writing a wrapper client class so that my protocol class won't need to be exposed to home-assistant directly. It will eliminate the need for home-assistant to handle protocol creation and destruction by handling that internally. I'll create a method in the wrapper client class which stops the automatic re-connection task.